rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+the foreseeable future. This section briefly discusses how this affects our
+compatibility guarantees.
+
+It is worth noting that among all these key dependencies, only TensorFlow has
+Artifact compatibility items, and everything else has only API compatibility.
+This is helpful for our effort, because we can do future major version upgrades
+(e.g, from Beam 2 to 3, if/when that happens) in a way hidden from our users, at
+least for all the first party components which we own.
+
+#### TensorFlow {#tensorflow}
+
+Because TensorFlow Transform (TFT) still relies on internal APIs of TensorFlow
+which is not subject to compatibility guarantees, each version of TFX libraries
+is only guaranteed to work with one version of TensorFlow.
+
+Meanwhile, there are users who need to use a different version of TensorFlow in
+their training code, therefore we anticipate a need to mix TFX components which
+link with different TensorFlow versions in the same pipeline. This is possible
+in a containerized deployment, because each component could be configured with
+different container images. In order to process Artifacts produced from
+different TensorFlow versions, we will rely on the guarantees provided by
+TensorFlow on SavedModel
+([source](https://www.tensorflow.org/guide/versions#compatibility_of_savedmodels_graphs_and_checkpoints)),
+most importantly:
+
+*   _In compliance with [semver](https://semver.org/), SavedModels written with
+    one version of TensorFlow can be loaded and evaluated with a later version
+    of TensorFlow with the same major release._
+*   _"Any SavedModel supported in TensorFlow major version N can be loaded and
+    executed with TensorFlow major version N+1"_
+
+Because we do not guarantee forward compatibility on our Artifacts payload, this
+means that if a component (i.e, _Trainer_) needs to use a different version of
+TensorFlow, then all _ancestor_ components which produced SavedModel (i.e,
+`Transform`) must use the same or older version of TFX.
+
+##### TensorFlow Enterprise 1.15 {#tensorflow-enterprise-1-15}
+
+The LTS version of TensorFlow Enterprise version poses a special challenge,
+because the 1.15 distribution is expected to be supported for three years
+starting from Oct 30, 2019. We propose to make the TFX 1.0 minor the only
+supported version for TensorFlow 1.15, and test this combination to the maximum
+extent possible. Any future versions of TFX 1.x will not be supported with TFE
+1.15.
+
+#### Apache Beam {#apache-beam}
+
+Apache Beam mostly follows semantic versioning: breakages are rare and we have
+only seen one occurrence related to type hints once in the entirety of 2020.
+Therefore we propose the following for all TFX 1.x versions:
+
+1.  Each version of TFX is guaranteed to be compatible with latest version of
+    `apache-beam` at release time;
+2.  In addition, `tfx`'s `install_requires` will allow future minor versions of
+    `apache-beam` in the same major version, i.e, `apache-beam>=2.27,<3`.
```

Comment:
TFX tends to stay up-to-date on our beam dependency, so that we can make sure latest optimizations and improvements from beam is used. My experience tells me that older version of beam usually works, but tends to fall into a under-tested scenario.

> cross language upgrades (JVM and Python)

Can you clarify this part a bit? My understanding is that we are requiring a python package `apache-beam` which by large does not force JVM part, and it generally supports most common Python versions, so I'm not exactly sure what you referred to here?

Reply:
Beam on 2.28 introduces the concept of cross language transforms. Due to our nature of monorepo (similar to Google's I believe) we only have a single version of Beam for our code, which means that we need to keep both java and Python based Beam dependencies in sync.

> My experience tells me that older version of beam usually works, but tends to fall into a under-tested scenario.

I agree, but setting it as a transtive dependency forces us to upgrade Beam if we want to upgrade TFX, which is largely undesired as it makes it hard.

Reply:
The primary concern from our side is how many old versions of Beam we have to support. Upgrading Beam versions would allow us to stay on top of latest improvements in Beam.

While a common answer would be "don't upgrade TFX until you are ready to upgrade both TFX and Beam", I understand you also want to use newer features in TFX.

I can see two high level paths from here:

Option 1 (slightly preferred by us): TFX will still use latest Beam, however we can try to open up all our release/packaging scripts and leave enough options for overriding Beam versions, so you can overwrite the version of beam in your own deployment. In this option, you are essentially depending on TFX on source and supporting yourself on the particular combination of TFX + Beam.  We still encourage you to upgrade Beam to as latest as possible. For this to work, I highly encourage you to use a containerized deployment.

Option 2 (not really preferred but with a precedence): we will try our best to make some versions of TFX to work with a (LTS) version of Beam. This is the special arrangement we are targeting with TensorFlow 1.15. Because there is a LTS version of that, we will try to make one version in TFX 1.x (likely 1.0) to work well enough with TensorFlow 1.15. This is mostly making sure there is a very stable combination which sees security backports, and I would not recommend users who wants to stay on top of latest features to use this. To make this work, we would need Beam team to commit to backport security fixes to this LTS version of Beam.

If any of this seems acceptable, I can go back to evaluate its concrete feasibility. WDYT?

Reply:
I think option 1 could work if we have a way to also run TFX end to end tests for the overwritten version of beam (to make sure its stable). 

Option 2 may be needed down the road, but this can be a separate discussion with Beam team.

Reply:
Will add following blurbs to the doc:


```
For users who wants to use older versions of `apache-beam` and `tfx` together,
while TFX team cannot directly support them in our released packages, we plan to
refactor our build scripts so that beam dependency can be overridden by
themselves, so they can build Python packages and container images in a
self-service fashion.
```